src/streams.ts

```diff
@@ -42,3 +55,116 @@ export function compose(streams: NodeJS.ReadWriteStream[]) {
     return new PassThrough({objectMode: true});
   }
 }
+
+/**
+ * A stream that accepts source file objects from a vinyl stream, and emits
+ * the file paths of all required file dependencies. Useful for building
+ * the dependency tree during build.
+ */
+export class SourceDependenciesStream extends Transform {
+
+  // A set of all dependency file paths passed through the stream
+  sourceGlobs: string[];
+  files = new Set<string>();
+
+  constructor(sourcGlobs: string[]) {
+    super({ objectMode: true });
+    this.sourceGlobs = sourcGlobs;
+  }
+
+  /**
+   * Process the given dependency before pushing it through the stream.
+   * Each dependency is only pushed through once to avoid duplicates.
+   */
+  pushDependency(dependencyPath: string) {
+    // If this is the first time we've seen this dependency, push it through.
+    if (this.files.has(dependencyPath)) {
+      return;
+    }
+
+    if (minimatchAll(dependencyPath, this.sourceGlobs)) {
+      console.log('MATCH SOURCE', dependencyPath);
+      return;
+    }
+
+    this.files.add(dependencyPath);
+    let dependency = new File({ path: dependencyPath });
+    this.push(dependency);
+  }
+
+  _transform(
+    file: File,
+    encoding: string,
+    callback: (error?: Error, data?: File) => void
+  ): void {
+    // If the file has already been seen, it was as the dependency of another
+    // source file. We can assume its dependencies were already parsed then.
+    if (this.files.has(file.path)) {
+      callback();
+      return;
+    }
+
+    // Only HTML source files can have dependencies. If the given file is not
+    // an HTML file, we can assume it has no dependencies.
+    if (file.extname !== '.html') {
+      callback();
+      return;
+    }
+
+    this._getDependencies(file.path).then((allDeps: DocumentDeps) => {
+      allDeps.imports.forEach(this.pushDependency.bind(this));
+      allDeps.scripts.forEach(this.pushDependency.bind(this));
+      allDeps.styles.forEach(this.pushDependency.bind(this));
+      callback();
+    }).catch(err => {
+      callback(err);
+    });
+  }
+
+  /**
+   * Create an analyzer and retreive all transitive dependencies for `url`
+   */
+  _getDependencies(url: string): Promise<DocumentDeps> {
+    let pathDirectory = posixPath.dirname(url);
+    let resolver = new FSResolver({});
+    let loader = new Loader();
+    loader.addResolver(resolver);
+    let analyzer = new Analyzer(false, loader);
```

Comment:
We already have an analyzer in the project, can't you just use that? Also, seems expensive to new up an Analyzer for each file.